### PR TITLE
zson.Unmarshal: Improve handling of null values

### PR DESCRIPTION
**Pull request by mattnibs**
This pr fixes a couple issues when unmarshaling null zed values:
- zson.Unmarshal was panicing when trying to unmarshal a "null" value
   into an array.
- Typed null values (e.g. "null(int64)") were not getting typed check if
   the go value was a pointer.
- Ensure nested pointers are getting properly traversed.

### CODE REVIEW

File: zson/marshal.go [hunk 1]
```diff
@@ -680,6 +680,13 @@ func (u *UnmarshalZNGContext) decodeAny(zv zed.Value, v reflect.Value) error {
 		v.Set(reflect.ValueOf(*zv.Copy()))
 		return nil
 	}
+	if typ := zed.TypeUnder(zv.Type); typ == zed.TypeNull && v.Kind() != reflect.Ptr {
+		// Since zed.TypeNull can be any type, set the interface to zero value.
+		// We need to make sure the reflect value isn't a pointer since pointers
+		// structs on null values should remain a nil pointer.
```
Comment (nwt) [inline zson/marshal.go:686]: I don't follow this. Can you elaborate on "pointers structs on null values should remain a nil pointer"?

File: zson/marshal_zng_test.go [hunk 2]
```diff
@@ -202,6 +202,31 @@ func TestUnmarshalRecord(t *testing.T) {
 	require.Equal(t, *v1.T1f1, *v3.T4f1)
 }
 
+func TestUnmarshalNull(t *testing.T) {
+	v, err := zson.ParseValue(zed.NewContext(), "null")
+	require.NoError(t, err)
```
Comment (nwt) [inline zson/marshal_zng_test.go:207]: Use `zed.Null`!

File: zson/marshal_zng_test.go [hunk 3]
```diff
@@ -202,6 +202,31 @@ func TestUnmarshalRecord(t *testing.T) {
...
+	var array []int
+	array = []int{1}
+	require.NoError(t, zson.UnmarshalZNG(*v, &array))
+	assert.Len(t, array, 0)
```
Comment (nwt) [inline zson/marshal_zng_test.go:211]:
```suggestion
	slice := []int{1}
	require.NoError(t, zson.UnmarshalZNG(*zed.Null, &slice))
	assert.Nil(t, slice)
```

File: zson/marshal_zng_test.go [hunk 4]
```diff
@@ -202,6 +202,31 @@ func TestUnmarshalRecord(t *testing.T) {
...
+	type testobj struct {
+		Val int
+	}
+	var obj *testobj
+	require.NoError(t, zson.UnmarshalZNG(*v, &obj))
+	assert.Nil(t, obj)
+	// Test if a non-array null returns an error.
+	v, err = zson.ParseValue(zed.NewContext(), "null(int64)")
+	require.NoError(t, err)
+	assert.EqualError(t, zson.UnmarshalZNG(*v, &array), "not an array")
```
Comment (nwt) [inline zson/marshal_zng_test.go:227]:
```suggestion
	assert.EqualError(t, zson.UnmarshalZNG(*zed.NullInt64, &slice), "not an array")
```

File: zson/marshal.go (hunk 1 above)
Comment (mattnibs) [reply zson/marshal.go:686]: Yeah I'm going to do another pass of comments on this logic because it is confusing.

File: zson/marshal_zng_test.go [hunk 5]
```diff
@@ -202,6 +202,28 @@ func TestUnmarshalRecord(t *testing.T) {
...
+	require.NoError(t, zson.UnmarshalZNG(*v, &integer))
+	assert.Equal(t, integer, 0)
+	m := map[string]string{"key": "value"}
+	require.NoError(t, zson.UnmarshalZNG(*v, &m))
+	assert.Nil(t, m)
+	type testobj struct {
+		Val int
+	}
+	var obj *testobj
+	require.NoError(t, zson.UnmarshalZNG(*v, &obj))
```
Comment (nwt) [inline zson/marshal_zng_test.go:221]:
```suggestion
	slice := []int{1}
	require.NoError(t, zson.UnmarshalZNG(*zed.Null, &slice))
	assert.Nil(t, slice)
	integer := -1
	require.NoError(t, zson.UnmarshalZNG(*zed.Null, &integer))
	assert.Equal(t, integer, 0)
	m := map[string]string{"key": "value"}
	require.NoError(t, zson.UnmarshalZNG(*zed.Null, &m))
	assert.Nil(t, m)
	type testobj struct {
		Val int
	}
	var obj *testobj
	require.NoError(t, zson.UnmarshalZNG(*zed.Null, &obj))
```

File: zson/marshal.go (hunk 1 above)
Comment (mattnibs) [reply zson/marshal.go]: Updated, I hope this is clearer.

File: zson/marshal_zng_test.go (hunk 5 above)
Comment (mattnibs) [reply zson/marshal_zng_test.go]: I reformatted this test.

**Comment by mccanne** [comment]
I added this to the end of TestUnmarshalNull() and it failed...
```
v, err = zson.ParseValue(zed.NewContext(), "null([int64])")
require.NoError(t, err)
slice = []int{1}
require.NoError(t, zson.UnmarshalZNG(*v, &slice))
assert.Nil(t, slice)
```

**Comment by mattnibs** [comment]
@mccanne yeah it wasn't doing inner type checking for maps, arrays, sets. I suppose we should do this here while we are at it.

**Comment by mattnibs** [comment]
@mccanne thanks I resolved the issue you found plus fixed the same issue for maps and bytes. Node: the code is not currently checking the inner types of null values- this doesn't seem trivial to do so I just put a comment in the code and perhaps revisit or I can fix now?

**Review by mccanne — APPROVED** [review]

**Pull request merged**